### PR TITLE
Multiple user back ends

### PR DESCRIPTION
**Pull request by schiessle**
add option to enable multiple user back-ends in parallel (see second check box from the saml admin settings) and to give the identity provider a custom display name (see first input field):

sso2

The user will then see this landing page and can decide between SSO and "direct login" which can be any user management which authenticates directly at Nextcloud (local user management, ldap,...). The SSO button has either the default name, as show here or the display name configured above.

sso2

@rullzer maybe you can give it a try and review it? Thanks!

### CODE REVIEW

**Comment by MorrisJobke** [comment]
I really like the idea. The only thing I see with this approach is that people need to know the difference somehow. Either customizable button labels (to suite the terms that are used for those login types in that environment) or some better generic terms (external and internal users - guests and members- etc) could help here to understand the difference from a nontechnical user point of view.

Maybe somebody from @nextcloud/designers has a good idea

**Comment by schiessle** [comment]
@MorrisJobke in a next step I want to introduce the possibility to have multiple SSO idp's then I want to add the possibility to give them custom names (and maybe logos). But I could already here allow to rename the saml button... let me check

**Comment by schiessle** [comment]
@MorrisJobke the button label is now configurable

File: templates/selectUserBackEnd.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<?php
+style('user_saml', 'selectUserBackEnd');
+
+/** @var array $_ */
+/** @var $l \OCP\IL10N */
+?>
+
+<div id="saml-select-user-back-end">
+
+<h1>Chose log in option:</h1>
```
Comment (blizzz) [inline templates/selectUserBackEnd.php]: typo

File: templates/selectUserBackEnd.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<?php
+style('user_saml', 'selectUserBackEnd');
+
+/** @var array $_ */
+/** @var $l \OCP\IL10N */
+?>
+
+<div id="saml-select-user-back-end">
+
+<h1>Chose login option:</h1>
```
Comment (blizzz) [inline templates/selectUserBackEnd.php]: typo is still here :) s/chose/choose

Comment (rullzer) [reply templates/selectUserBackEnd.php]: Ah sorry missed that...

Comment (schiessle) [reply templates/selectUserBackEnd.php]: no problem... will fix the typo quickly

File: lib/Controller/SAMLController.php [hunk 3]
```diff
@@ -309,4 +309,77 @@ public function genericError($message) {
...
+	 */
+	private function getSSOUrl($redirectUrl) {
+
+		$originalUrl = '';
+		if(!empty($redirectUrl)) {
+			$originalUrl = $this->urlGenerator->getAbsoluteURL($redirectUrl);
+		}
+
+
+		$csrfToken = \OC::$server->getCsrfTokenManager()->getToken();
```
Comment (MorrisJobke) [inline lib/Controller/SAMLController.php]: Should be injected as dependency in the constructor.

**Pull request merged**